File: 002/002.scala
object Main{
  //scala Main  0.78s user 0.10s system 122% cpu 0.714 total

  def main(args: Array[String]): Unit = {
    println(fibsLessThanN(4000000).filter(even).sum)
  }

  def fibsLessThanN(n: Long): List[Long] = {
    fibsLessThanNPrime(n, 0, 1, List())
  }

  def fibsLessThanNPrime(n: Long, prev: Long, curr: Long, acc: List[Long]): List[Long] = {
    if(curr > n) acc else fibsLessThanNPrime(n, curr, prev + curr, curr :: acc)
  }

  def even(n: Long) = {
    n % 2 == 0
  }
}

File: 002/002.rs
/*
./rs  0.00s user 0.00s system 0% cpu 0.004 total
*/

fn main(){
  println!("{}", fib_sum(4_000_000));
}

fn fib_sum(i : u64) -> u64{
    let (mut a, mut b, mut sum) = (1, 1, 0);
    while a < i {
        if a & 1 != 0 {sum = sum + a;};
        b = a + b;
        a = b - a;
    }
    sum
}
